File: https:/github.com/abhinavtripathi17/DSA-Leetcode.git/61-RotateList/61-RotateList.java
// Last updated: 8/4/2026, 7:24:07 PM
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if(head == null || head.next == null || k == 0) return head;
        ListNode res = head;

        int cnt = 1;

        while(res.next != null){
            cnt++;
            res = res.next;
        }
        k = k % cnt;

        if(k == 0) return head;

        res.next = head; // ll is circular now

        int x = cnt - k;

        //res.next = head;
        while(x > 0){
            res = res.next;
            x--;
        }
        head = res.next;
        res.next = null;

        return head;
    }
}

File: https:/github.com/abhinavtripathi17/DSA-Leetcode.git/61-RotateList/61-RotateList.cpp
// Last updated: 8/4/2026, 7:24:01 PM
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(head == nullptr || head->next == nullptr) return head;
        ListNode* temp = head;
        ListNode* prev = head;
        int cnt = 0;
        while(temp != nullptr){
            prev = temp;
            temp = temp->next;
            cnt++;
        }
        prev->next = head;

       k = k%cnt;
       int len = cnt - k;

       temp = head;
       prev = temp;
       while(len > 0){
        prev = temp;
        temp = temp->next;
        len--;
       }
       prev->next = nullptr;

       return temp;
    }
};